src/webgpu/api/operation/memory_sync/buffer/buffer_sync_test.ts

```diff
@@ -0,0 +1,242 @@
+
+    return this.device.createRenderPipeline({
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'vert_main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'frag_main',
+      },
+      primitiveTopology: 'point-list',
+      colorStates: [{ format: 'rgba8unorm' }],
+    });
+  }
+
+  beginSimpleRenderPass(encoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const view = this.device
+      .createTexture({
+        size: { width: 4, height: 4, depth: 1 },
+        format: 'rgba8unorm',
+        usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+      })
+      .createView();
+    return encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: view,
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+        },
+      ],
+    });
+  }
+
+  // Write buffer via draw call in render pass. Use bundle if needed.
+  writeByRenderPass(bundle: boolean, buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteRenderPipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+
+    const pass = this.beginSimpleRenderPass(encoder);
+    const renderer = bundle
+      ? this.device.createRenderBundleEncoder({ colorFormats: ['rgba8unorm'] })
+      : pass;
+    renderer.setBindGroup(0, bindGroup);
+    renderer.setPipeline(pipeline);
+    renderer.draw(1, 1, 0, 0);
+
+    if (bundle) pass.executeBundles([(renderer as GPURenderBundleEncoder).finish()]);
+    pass.endPass();
+  }
+
+  // Write buffer via dispatch call in compute pass.
+  writeByComputePass(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteComputePipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatch(1);
+    pass.endPass();
+  }
+
+  // Write bufer via BuferToBuffer copy.
+  async writeByB2BCopy(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const tmpBuffer = await this.createBufferWithValue(value);
+    encoder.copyBufferToBuffer(tmpBuffer, 0, buffer, 0, SIZE);
+  }
+
+  // Write buffer via TextureToBuffer copy.
+  async writeByT2BCopy(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const tmpBuffer = await this.createBufferWithValue(value);
+    const tmpTexture = this.device.createTexture({
+      size: { width: 1, height: 1, depth: 1 },
+      format: 'rgba8uint',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+    encoder.copyBufferToTexture(
+      { buffer: tmpBuffer, bytesPerRow: 256 },
+      { texture: tmpTexture, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
+      { width: 1, height: 1, depth: 1 }
+    );
```

Comment:
The initialization of the source texture should ideally be separated from the test copyTextureToBuffer call by a fence, like is done for the buffer initialization. Instead of `await createBufferWithValue`+`createTexture`+`copyBufferToTexture`, consider `createTexture`+`writeTexture`+`signal`+`await onCompletion`.

Reply:
Done.